tempesta_fw/http.c

```diff
@@ -615,35 +634,212 @@ tfw_http_adjust_resp(TfwHttpResp *resp, TfwHttpReq *req)
 	return tfw_http_set_hdr_connection(m, conn_flg);
 }
 
+static inline bool
+tfw_http_req_is_idempotent(TfwHttpReq *req)
+{
+	return (req->method != TFW_HTTP_METH_POST);
+}
+
 /*
- * Depending on results of processing of a request, either send the request
- * to an appropriate server, or return the cached response. If none of that
- * can be done for any reason, return HTTP 404 or 500 error to the client.
+ * Forward request @req to server connection @srv_conn.
+ * Can be called in both SoftIRQ and kernel thread contexts.
+ */
+static void
+tfw_http_conn_req_forward(TfwConnection *srv_conn, TfwHttpReq *req)
+{
+	TFW_DBG3("%s: srv_conn = %p, req = %p\n", __func__, srv_conn, req);
+	BUG_ON(!(TFW_CONN_TYPE(srv_conn) & Conn_Srv));
+
+	spin_lock_bh(&srv_conn->msg_qlock);
+	if (!tfw_http_req_is_idempotent(req))
+		req->flags |= TFW_HTTP_SYNC_REQ;
+	list_add_tail(&req->msg.fwd_list, &srv_conn->msg_queue);
+	if (srv_conn->flags & TFW_CONN_FWD_HOLD) {
+		spin_unlock_bh(&srv_conn->msg_qlock);
+		return;
```

Comment:
Hm, just return? So `tfw_http_req_cache_cb()` now ends with assumption that the request was sent to server, however it actually wasn't. Who and when sends the request now?